### ISSUE TITLE
topology_custom.test_replace_ignore_nodes is flaky

### ISSUE THREAD

**Issue by michoecho** [scylladb/scylladb#14646]
The scenario, as far as I understand it:

The test kills nodes `127.72.81.12`, `127.72.81.32` and gracefully stops `127.72.81.48`.

Then it replaces `127.72.81.12` with `127.72.81.35` using `ignore_dead_nodes=[127.72.81.32, 127.72.81.48]`. This succeeds.

Then it replaces `127.72.81.32` with `127.72.81.57` using `ignore_dead_nodes=[127.72.81.48]`. This is expected to succeed, but fails, because `127.72.81.57`'s gossip refuses to start without `127.72.81.12` present:
```
init - Startup failed: std::runtime_error (Failed to mark node as alive in 30000 ms, nodes={127.72.81.57, 127.72.81.35, 127.72.81.43, 127.72.81.41, 127.72.81.39, 127.72.81.12, 127.72.81.23}, live_nodes={127.72.81.57, 127.72.81.35, 127.72.81.43, 127.72.81.41, 127.72.81.39, 127.72.81.23})
```
sources:
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2404/artifact/testlog/x86_64/dev/scylla-180.log
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2404/artifact/testlog/x86_64/dev/topology_custom.test_replace_ignore_nodes.1.log

So IIUC the problem is that `127.72.81.57` didn't get the memo that `127.72.81.12` is gone and shouldn't be waited for, even though its has already been replaced before `127.72.81.57` is started.

**Comment by michoecho** [comment]
@kbr-scylla You added this test a few days ago in https://github.com/scylladb/scylladb/pull/14507.

**Comment by kbr-scylla** [comment]
> So IIUC the problem is that 127.72.81.57 didn't get the memo that 127.72.81.12 is gone and shouldn't be waited for, even though its has already been replaced before 127.72.81.12 is started.

Hm, actually it did:
```
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -4951537701902447216. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -6108029415812014824. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -6907664429548326176. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -5001876849851725997. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -6511544901280451354. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 4202010258256783415. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -3106559103669268007. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 4221809952587115440. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,827 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -6492096531581779188. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 7776513095124287641. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 333039822775427100. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 146434312805889274. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 7048689242400418103. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 6633852014283034353. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token -4861963604709447521. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: Nodes 127.72.81.12 and 127.72.81.35 have the same token 4976079165281076413. Ignoring 127.72.81.12
INFO  2023-07-11 21:50:27,828 [shard 0] storage_service - handle_state_normal: endpoints_to_remove endpoint=127.72.81.12
```

Which means that `token_metadata`/`topology` should no longer contain `127.72.81.12`, so the calculation of `sync_nodes`:
```cpp
        std::vector<gms::inet_address> sync_nodes;
        tm.get_topology().for_each_node([&] (const locator::node* np) {
            auto ep = np->endpoint();
            if (!ignore_nodes.contains(ep) && (!ri || ep != ri->address)) {
                sync_nodes.push_back(ep);
            }
        });
```

should not observe this node.

cc @bhalevy

**Comment by bhalevy** [comment]
topology may contain nodes that have no tokens assigned to them.
we should check `is_normal_token_owner` for each node in topology in this case.
or walk `tm.get_all_endpoints()` (which actually returns a reference to `_normal_token_owners`)

**Comment by kbr-scylla** [comment]
For correctness, the node must not appear in `_normal_state_handled_on_boot` until `remove_endpoint` -- and all the background work that it starts -- has finished.

cc @asias

**Comment by kbr-scylla** [comment]
Is this comment even relevant today?
```cpp
    // We can not run on_remove callbacks here becasue on_remove in
    // storage_service might take the gossiper::timer_callback_lock
```

There is no such thing as `gossiper::timer_callback_lock` and there are no gossiper-specific locks that `storage_service::on_remove` would take. It's taking the same locks as in `handle_state_normal` and it's doing a subset of its operations.

Maybe we can put the loop into foreground?

**Comment by kbr-scylla** [comment]
> It's taking the same locks as in handle_state_normal

-- and it's releasing them before calling `remove_endpoint`

**Comment by kbr-scylla** [comment]
```
    // do subscribers first so anything in the subscriber that depends on gossiper state won't get confused
```
It's funny that the comment says "we need to do this operation *first* so things don't break"... and then it does the operation in the background...

**Comment by bhalevy** [comment]
I think it's a good direction.
@asias, how about you?

**Comment by kbr-scylla** [comment]
Proposed fix: https://github.com/kbr-scylla/scylladb/commit/6d9f45f02acb2cb939cbfe4feceee73b70711553
Started dtests `update_cluster_layout_tests.py replace_address_test.py repair_based_node_operations_test.py -m 'dtest_full and not dtest_long'` on this commit: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1807/

**Comment by asias** [comment]
> Proposed fix: [kbr-scylla@6d9f45f](https://github.com/kbr-scylla/scylladb/commit/6d9f45f02acb2cb939cbfe4feceee73b70711553) Started dtests `update_cluster_layout_tests.py replace_address_test.py repair_based_node_operations_test.py -m 'dtest_full and not dtest_long'` on this commit: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1807/
> 
> In-tree tests are passing. (In dev mode.)

let's ref the commit 19e8c14ad1a1c79a0432dd4b3b3b84c1ba36927b which get rid of the timer_callback_lock.

**Comment by asias** [comment]
> I think it's a good direction. @asias, how about you?

Looks good to me.

**Comment by avikivity** [comment]
@kbr-scylla please advise re backport

Seems like a change that could indirectly depend on other refactoring in gossip, so risky.

**Comment by kbr-scylla** [comment]
The test that was flaky and fixed by this is not in 5.2 or earlier branches, so no need to backport.
It's still a valuable change, but not critical.
